### PR TITLE
Add ReportAnalyzer utility class

### PR DESCRIPTION
**Pull request by AlessandroPierro**
Objective of pull request: introduce an utility class to analyze and plot the results of OptimizationSolver

## Pull request checklist

Your PR fulfills the following requirements:
- [x] Tests are part of the PR (for bug fixes / features)
- [x] [Docs](https://github.com/lava-nc/docs) reviewed and added / updated if needed (for bug fixes / features)
- [x] PR conforms to [Coding Conventions](https://lava-nc.org/developer_guide.html#coding-conventions)
- [x] [PR applys BSD 3-clause or LGPL2.1+ Licenses](https://lava-nc.org/developer_guide.html#add-a-license) to all code files
- [x] Lint (`pyb`) passes locally
- [x] Build tests (`pyb -E unit`) or (`python -m unittest`) passes locally

## Pull request type

 

 

Please check your PR type:
- [x] Feature

## What is the new behavior?

-

## Does this introduce a breaking change?

- [x] No

### CODE REVIEW

File: src/lava/lib/optimization/utils/state_analysis.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analysis.py:1]: Thanks Alessandro, 
This looks really good,!

The only thing we have misalignment is in initializing the class with problem and report. I had in mind initializing it only with the problem, and passing the report to the methods. This is because the problem defines some persistent components like, dimension of the state space, or global minimum, in contrast, different reports could be used over the same problem/state space. With that rationale, my advice would be to accept the report in the methods.

Recall we have things like compute cost on the problem definition, think also about what responsibilities are of the analysis tool and which ones of the problem class. For example, which class should produce the state space, or a sample set over it?

**Review by GaboFGuerra — COMMENTED** [review]
Great progress!

File: src/lava/lib/optimization/utils/state_analysis.py
Comment (GaboFGuerra) [reply src/lava/lib/optimization/utils/state_analysis.py:1]: Forgot to mention, my suggestion is actually aligned with the fact that you are already extracting cost and state every time the public methods are called.

Comment (AlessandroPierro) [reply src/lava/lib/optimization/utils/state_analysis.py:1]: Thanks for the feedback, I understood your point of view and implemented the changes.

**Comment by AlessandroPierro** [comment]
Ready for review! 🎉

The tests are passing locally and will pass on the CI once #222 is merged.

Here it is a preview of plots generated with StateAnalyzer. 

Your feedback would be greatly appreciated! ❤

plot_cost_distribution
plot_cost_timeseries
plot_delta_cost_distribution
plot_min_cost_timseries
plot_num_visited_states
plot_state_analysis_summary
plot_state_timeseries
plot_successive_states_distance

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analyzer.py:21]: plot instead of plots

Comment (GaboFGuerra) [reply src/lava/lib/optimization/utils/state_analyzer.py:21]: Maybe also phrase it in a way that explains the presence of "problem" as input parameter  in the __init__ method

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analyzer.py:27]: I think you only have to add docstring to either the class or the __init__ method.

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analyzer.py:32]: are we analyzing a problem or a solver?

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analyzer.py:35]: the following four methods are missing docstrings.

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analyzer.py:101]: arguments not documented in docstring of this an other methods below.

Comment (GaboFGuerra) [reply src/lava/lib/optimization/utils/state_analyzer.py:101]: would phrase "as a function of" rather than "through"

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analyzer.py:126]: method is min_cost but ylabel is Best Cost.

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analyzer.py:141]: what about plot_cost_transitions_distribution ?

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analyzer.py:176]: name is unclear for outsiders. What about plot_state_transition_distance or even plot_states_hamming_distance

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (GaboFGuerra) [inline src/lava/lib/optimization/utils/state_analyzer.py:1]: Many methods have missing or incomplete docstrings.

File: tests/lava/lib/optimization/utils/test_state_analyzer.py
Comment (GaboFGuerra) [inline tests/lava/lib/optimization/utils/test_state_analyzer.py:1]: why are the remove files commented out? This is important to avoid accumulation of clutter. You can have it as a tearDown method, similar to the setUp one you are already using. tearDown will execute after each individual test.

Comment (GaboFGuerra) [inline tests/lava/lib/optimization/utils/test_state_analyzer.py:1]: The tests only test the existence of the file. Is there anything else that needs to be tested?

**Review by GaboFGuerra — APPROVED** [review]
Great Tool! Thanks Alessandro, a few minor suggestions in the comments.

**Review by GaboFGuerra — APPROVED** [review]
Great Tool! Thanks Alessandro, a few minor suggestions in the comments.

File: src/lava/lib/optimization/utils/state_analyzer.py
Comment (AlessandroPierro) [reply src/lava/lib/optimization/utils/state_analyzer.py:176]: Nice suggestion!

**Comment by mgkwill** [comment]
Hi guys. I don't see "StateAnalyser" added as a dependency in this change. What is the version added? Please add to pyproject.toml and update poetry.lock.

**Comment by mgkwill** [comment]
> Hi guys. I don't see "StateAnalyser" added as a dependency in this change. What is the version added? Please add to pyproject.toml and update poetry.lock.

Is it just seaborn that needs to be added to deps for approval?

**Comment by AlessandroPierro** [comment]
> > Hi guys. I don't see "StateAnalyser" added as a dependency in this change. What is the version added? Please add to pyproject.toml and update poetry.lock.
> 
> Is it just seaborn that needs to be added to deps for approval?

Indeed, we just need approval for seaborn. Thanks for looking into that!

**Pull request merged**
